File: src/lib/community-recommendations.test.ts
import { describe, expect, it } from "vitest";
import {
  cohortConfidence,
  confidenceLevel,
  deriveRecommendations,
  inventoryBrandLink,
  sourceCompsLink,
} from "./community-recommendations";
import type { CommunityBenchmarks } from "./community-benchmarks";

function benchmarks(
  partial: Partial<CommunityBenchmarks>,
): CommunityBenchmarks {
  return {
    meta: { minSellers: 5, periodStart: null, generatedAt: "2026-06-18" },
    topBrands: [],
    categories: [],
    trendingCategories: [],
    priceRealization: null,
    timeToSell: null,
    trends: { windows: { d30: null, d90: null, d365: null }, monthly: [] },
    you: {
      listed: 0,
      sold: 0,
      sellThrough: null,
      medianDaysToSell: null,
      medianRealization: null,
      peerComparison: null,
    },
    ...partial,
  };
}

describe("cohortConfidence", () => {
  it("floors at the k-anonymity threshold and climbs with sellers", () => {
    expect(cohortConfidence(5)).toBeCloseTo(0.4, 5);
    expect(cohortConfidence(20)).toBeCloseTo(0.85, 5);
    // Caps at 0.95 no matter how large the cohort.
    expect(cohortConfidence(100)).toBe(0.95);
  });

  it("never returns below the floor for tiny cohorts", () => {
    expect(cohortConfidence(0)).toBe(0.4);
  });
});

describe("confidenceLevel", () => {
  it("buckets into high/medium/low", () => {
    expect(confidenceLevel(0.8)).toBe("high");
    expect(confidenceLevel(0.6)).toBe("medium");
    expect(confidenceLevel(0.5)).toBe("low");
  });
});

describe("deriveRecommendations", () => {
  it("suggests sourcing a high-sell-through brand and pricing it near the avg", () => {
    const recs = deriveRecommendations(
      benchmarks({
        topBrands: [
          {
            brand: "Patagonia",
            sellers: 20,
            listed: 100,
            sold: 80,
            sellThrough: 0.8,
            avgSalePrice: 65,
          },
        ],
      }),
    );
    const source = recs.find((r) => r.kind === "source");
    const price = recs.find((r) => r.kind === "price");
    expect(source).toBeDefined();
    expect(source?.subject).toBe("Patagonia");
    expect(source?.cohortSize).toBe(20);
    expect(source?.deepLink).toBe(sourceCompsLink("Patagonia"));
    expect(price).toBeDefined();
    expect(price?.title).toContain("$65.00");
    expect(price?.deepLink).toBe(inventoryBrandLink("Patagonia"));
  });

  it("does NOT suggest sourcing a low-sell-through brand", () => {
    const recs = deriveRecommendations(
      benchmarks({
        topBrands: [
          {
            brand: "Slowmover",
            sellers: 12,
            listed: 100,
            sold: 20,
            sellThrough: 0.2,
            avgSalePrice: 30,
          },
        ],
      }),
    );
    expect(recs.some((r) => r.kind === "source")).toBe(false);
    // ...but a price reference is still useful.
    expect(recs.some((r) => r.kind === "price")).toBe(true);
  });

  it("surfaces a rising category and skips flat/declining ones", () => {
    const recs = deriveRecommendations(
      benchmarks({
        trendingCategories: [
          {
            category: "outerwear",
            sellers: 15,
            soldRecent: 60,
            soldPrevious: 30,
            growth: 1.0,
          },
          {
            category: "tops",
            sellers: 10,
            soldRecent: 31,
            soldPrevious: 30,
            growth: 0.03,
          },
        ],
      }),
    );
    const cats = recs.filter((r) => r.subject === "outerwear");
    expect(cats).toHaveLength(1);
    expect(recs.some((r) => r.subject === "tops")).toBe(false);
  });

  it("ranks by score (strong signal + large cohort first)", () => {
    const recs = deriveRecommendations(
      benchmarks({
        topBrands: [
          {
            brand: "Weak",
            sellers: 5,
            listed: 10,
            sold: 6,
            sellThrough: 0.6,
            avgSalePrice: 20,
          },
          {
            brand: "Strong",
            sellers: 25,
            listed: 100,
            sold: 95,
            sellThrough: 0.95,
            avgSalePrice: 80,
          },
        ],
      }),
    );
    expect(recs[0]?.subject).toBe("Strong");
  });

  it("returns nothing when there are no qualifying cohorts", () => {
    expect(deriveRecommendations(benchmarks({}))).toEqual([]);
  });
});

File: android/app/src/test/java/com/gradethread/app/analytics/CommunityRecommendationsTest.kt
package com.gradethread.app.analytics

import com.gradethread.app.inventory.InventoryFilterRequests
import org.junit.After
import com.gradethread.app.R
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * US-1369: the recommendation thresholds, the confidence model, and the states
 * that look the same on screen but mean different things.
 */
class CommunityRecommendationsTest {

    @After
    fun tearDown() = InventoryFilterRequests.clear()

    private fun brand(
        name: String = "Patagonia",
        sellers: Int = 10,
        listed: Int = 100,
        sold: Int = 60,
        sellThrough: Double? = 0.6,
        avgSalePrice: Double? = 85.0,
    ) = BrandBenchmark(name, sellers, listed, sold, sellThrough, avgSalePrice)

    private fun category(name: String = "Outerwear", sellers: Int = 8, growth: Double? = 0.25) =
        CategoryTrend(name, sellers, soldRecent = 50, soldPrevious = 40, growth = growth)

    // ── Confidence ───────────────────────────────────────────────────────────

    @Test
    fun `confidence starts at the k-anonymity floor and is capped below certainty`() {
        // A cohort is never certainty. A "100% confidence" badge on somebody
        // else's aggregate would be a promise this data cannot make.
        assertEquals(0.40, CommunityRecommendations.cohortConfidence(5), 0.0001)
        assertEquals(0.55, CommunityRecommendations.cohortConfidence(10), 0.0001)
        assertEquals(0.95, CommunityRecommendations.cohortConfidence(1000), 0.0001)
        // Below the floor can't reach the client at all, but it must not read
        // as MORE confident than the floor either.
        assertEquals(0.40, CommunityRecommendations.cohortConfidence(1), 0.0001)
    }

    @Test
    fun `confidence bands match the labels`() {
        assertEquals(ConfidenceLevel.HIGH, CommunityRecommendations.confidenceLevel(0.75))
        assertEquals(ConfidenceLevel.MEDIUM, CommunityRecommendations.confidenceLevel(0.55))
        assertEquals(ConfidenceLevel.LOW, CommunityRecommendations.confidenceLevel(0.54))
    }

    // ── Thresholds ───────────────────────────────────────────────────────────

    @Test
    fun `a brand below the sell-through floor is not suggested for sourcing`() {
        val weak = CommunityRecommendations.derive(
            CommunityBenchmarks(topBrands = listOf(brand(sellThrough = 0.49))),
        )
        assertTrue(weak.none { it.kind == RecommendationKind.SOURCE })
        // The pricing suggestion still stands: knowing what a brand sells FOR
        // is useful even when it sells slowly.
        assertTrue(weak.any { it.kind == RecommendationKind.PRICE })
    }

    @Test
    fun `a brand with no sell-through reading gets no sourcing advice`() {
        val recs = CommunityRecommendations.derive(
            CommunityBenchmarks(topBrands = listOf(brand(sellThrough = null))),
        )
        assertTrue(recs.none { it.kind == RecommendationKind.SOURCE })
    }

    @Test
    fun `a brand with no average price gets no pricing advice`() {
        val recs = CommunityRecommendations.derive(
            CommunityBenchmarks(topBrands = listOf(brand(avgSalePrice = null))),
        )
        assertTrue(recs.none { it.kind == RecommendationKind.PRICE })
        assertTrue(recs.any { it.kind == RecommendationKind.SOURCE })
    }

    @Test
    fun `a zero average price is not a price to aim at`() {
        val recs = CommunityRecommendations.derive(
            CommunityBenchmarks(topBrands = listOf(brand(avgSalePrice = 0.0))),
        )
        assertTrue(recs.none { it.kind == RecommendationKind.PRICE })
    }

    @Test
    fun `a category below the growth floor is not called rising`() {
        assertTrue(
            CommunityRecommendations.derive(
                CommunityBenchmarks(trendingCategories = listOf(category(growth = 0.14))),
            ).isEmpty(),
        )
        assertEquals(
            1,
            CommunityRecommendations.derive(
                CommunityBenchmarks(trendingCategories = listOf(category(growth = 0.15))),
            ).size,
        )
    }

    // ── Ranking and content ──────────────────────────────────────────────────

    @Test
    fun `recommendations are ranked best first`() {
        val recs = CommunityRecommendations.derive(
            CommunityBenchmarks(
                topBrands = listOf(
                    brand("Slow", sellThrough = 0.51, avgSalePrice = 20.0),
                    brand("Fast", sellThrough = 0.95, avgSalePrice = 20.0),
                ),
            ),
        )
        val sources = recs.filter { it.kind == RecommendationKind.SOURCE }
        assertEquals(listOf("Fast", "Slow"), sources.map { it.subject })
        assertTrue(recs.zipWithNext().all { (a, b) -> a.score >= b.score })
    }

    @Test
    fun `a brand recommendation carries the filter to deep-link with`() {
        val rec = CommunityRecommendations.derive(
            CommunityBenchmarks(topBrands = listOf(brand("Carhartt"))),
        ).first()

        assertEquals("Carhartt", rec.brandFilter)
        // US-2976: the COHORT SIZE is what "sellers" was standing in for - the
        // number a seller weighs the recommendation by. It is the first
        // argument in every detail shape, and it picks the plural form.
        //
        // Deliberately NOT asserting which resource: `first()` is whichever
        // recommendation ranks highest, and pinning that here would make this
        // test fail whenever the SCORING changes, which is not what it is for.
        assertEquals(rec.cohortSize, rec.detail.args[0])
        assertEquals(rec.cohortSize, rec.detail.quantity)
    }

    @Test
    fun `a category recommendation has nothing to filter by`() {
        // The local item mirror has no category column, so a category filter
        // would match nothing and the tap would look broken.
        val rec = CommunityRecommendations.derive(
            CommunityBenchmarks(trendingCategories = listOf(category("Denim"))),
        ).single()

        assertNull(rec.brandFilter)
        assertEquals("Denim", rec.subject)
    }

    // ── Empty versus nothing-actionable ──────────────────────────────────────

    @Test
    fun `having data and having advice are different states`() {
        // These look identical on screen and mean opposite things: one is a
        // wait, the other is a judgement.
        val quiet = CommunityBenchmarks(topBrands = listOf(brand(sellThrough = 0.1, avgSalePrice = null)))
        assertTrue(quiet.hasBenchmarkData)
        assertTrue(CommunityRecommendations.derive(quiet).isEmpty())

        assertFalse(CommunityBenchmarks().hasBenchmarkData)
    }

    // ── Peer standing ────────────────────────────────────────────────────────

    @Test
    fun `peer standing names where you sit`() {
        val you = SellerSummary(
            listed = 40,
            sold = 30,
            sellThrough = 0.75,
            peerComparison = PeerComparison(
                peerCount = 12,
                peerMedianSellThrough = 0.5,
                yourSellThrough = 0.75,
                percentile = 0.8,
            ),
        )
        val text = CommunityRecommendations.peerStanding(you)!!

        assertTrue(text.contains("75%"))
        assertTrue(text.contains("above"))
        assertTrue(text.contains("12 sellers"))
        assertNull(CommunityRecommendations.peerStandingBlocker(you))
    }

    @Test
    fun `no comparison is withheld rather than invented`() {
        // The RPC omits peerComparison below five peers. A percentile with no
        // rate behind it is not a position.
        assertNull(CommunityRecommendations.peerStanding(SellerSummary(listed = 40)))
        assertNull(
            CommunityRecommendations.peerStanding(
                SellerSummary(
                    listed = 40,
                    peerComparison = PeerComparison(peerCount = 9, yourSellThrough = null),
                ),
            ),
        )
    }

    @Test
    fun `a thin seller is told what would unlock the comparison`() {
        // Two DIFFERENT resources: one names a thing the seller can do, the
        // other says to wait. Telling them apart is the whole point.
        assertEquals(
            R.string.community_need_three,
            CommunityRecommendations.peerStandingBlocker(SellerSummary(listed = 1)),
        )
        assertEquals(
            R.string.community_no_peers,
            CommunityRecommendations.peerStandingBlocker(SellerSummary(listed = 40)),
        )
    }

    // ── Deep-link handoff ────────────────────────────────────────────────────

    @Test
    fun `a brand request is consumed exactly once`() {
        // Left set, the filter would re-apply every time the seller returned to
        // the tab and feel impossible to clear.
        InventoryFilterRequests.requestBrand("Patagonia")
        assertEquals("Patagonia", InventoryFilterRequests.consumeBrand())
        assertNull(InventoryFilterRequests.consumeBrand())
    }

    @Test
    fun `a blank brand is never requested`() {
        InventoryFilterRequests.requestBrand("   ")
        assertNull(InventoryFilterRequests.brand.value)
    }

    @Test
    fun `a brand request is trimmed`() {
        InventoryFilterRequests.requestBrand("  Nike ")
        assertEquals("Nike", InventoryFilterRequests.brand.value)
        assertNotNull(InventoryFilterRequests.consumeBrand())
    }
}
